internal/topology/node.go

```diff
@@ -0,0 +1,218 @@
+package topology
+
+import (
+	"context"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/contentsquare/chproxy/internal/counter"
+	"github.com/contentsquare/chproxy/internal/heartbeat"
+	"github.com/contentsquare/chproxy/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// prevents excess goroutine creating while penalizing overloaded host
+	DefaultPenaltySize     = 5
+	DefaultMaxSize         = 300
+	DefaultPenaltyDuration = time.Second * 10
+)
+
+type nodeOpts struct {
+	defaultActive   bool
+	penaltySize     uint32
+	penaltyMaxSize  uint32
+	penaltyDuration time.Duration
+}
+
+func defaultNodeOpts() nodeOpts {
+	return nodeOpts{
+		penaltySize:     DefaultPenaltySize,
+		penaltyMaxSize:  DefaultMaxSize,
+		penaltyDuration: DefaultPenaltyDuration,
+	}
+}
+
+type NodeOption interface {
+	apply(*nodeOpts)
+}
+
+type penalty struct {
+	penaltySize     uint32
+	penaltyMaxSize  uint32
+	penaltyDuration time.Duration
+}
+
+func (o penalty) apply(opts *nodeOpts) {
+	opts.penaltySize = o.penaltySize
+	opts.penaltyMaxSize = o.penaltyMaxSize
+	opts.penaltyDuration = o.penaltyDuration
+}
+
+func WithPenalty(size, max uint32, duration time.Duration) NodeOption {
```

Comment:
I've included some opts, for now they will only be useful in tests. In the future it could be interesting in case users want to tweak e.g. the penalty for failure and have fine-grained control over the load balancing of chproxy.